src/grab.ts

```diff
@@ -1,157 +1,123 @@
-          } else {
-            acc[key] = value;
-          }
-          return acc;
-        }, {});
+  const newSchema = (() => {
+    const toSchemaInput = (sel: Selection) =>
+      Object.entries(sel).reduce<z.ZodRawShape>((acc, [key, value]) => {
+        if (value instanceof BaseResult) {
+          acc[key] = value.schema;
+        } else if (Array.isArray(value)) {
+          acc[key] = value[1];
+        } else if (value instanceof z.ZodType) {
+          acc[key] = value;
+        }
+        return acc;
+      }, {});
 
-      // Unknown schema means we just use the selection passed
-      if (
-        (prev.schema instanceof z.ZodArray
-          ? prev.schema.element
-          : prev.schema) instanceof z.ZodUnknown
-      ) {
-        // Split base and conditional fields
-        const conditionalFields = Object.values(
-          conditionalSelections || {}
-        ) as Selection[];
+    // Split base and conditional fields
+    const conditionalFields = Object.values(
+      conditionalSelections || {}
+    ) as Selection[];
 
-        const baseSchema = z.object(toSchemaInput(selection));
-        const conditionalFieldSchemas = conditionalFields.map((field) =>
-          baseSchema.merge(z.object(toSchemaInput(field)))
-        );
-        const s =
-          conditionalFieldSchemas.length === 0
-            ? baseSchema
-            : // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-              // @ts-ignore Need to figure out how to make tsc happy
-              z.union([...conditionalFieldSchemas, baseSchema]);
+    const baseSchema = z.object(toSchemaInput(selection));
+    const conditionalFieldSchemas = conditionalFields.map((field) =>
+      baseSchema.merge(z.object(toSchemaInput(field)))
+    );
+    const s =
+      conditionalFieldSchemas.length === 0
+        ? baseSchema
+        : // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore Need to figure out how to make tsc happy
+          z.union([...conditionalFieldSchemas, baseSchema]);
 
-        return prev.schema instanceof z.ZodArray ? z.array(s) : s;
-      }
-      // If we're already dealing with an object schema inside our array, we need to do a Pick
-      else if (
-        (prev.schema instanceof z.ZodArray
-          ? prev.schema.element
-          : prev.schema) instanceof z.ZodObject
-      ) {
-        const toPick = Object.keys(selection).reduce<{
-          [key: string]: true;
-        }>((acc, key) => {
-          acc[key] = true;
-          return acc;
-        }, {});
+    return schema instanceof z.ZodArray ? z.array(s) : s;
+  })();
 
-        return prev.schema instanceof z.ZodArray
-          ? z.array(prev.schema.element.pick(toPick))
-          : (prev.schema as z.ZodObject<any>).pick(toPick);
-      }
-      // If not unknown/object, I don't know what happened 👀
-      else {
-        return z.never();
-      }
-    })();
+  const res = (schema instanceof z.ZodArray
```

Comment:
```suggestion
  const res = (newSchema instanceof z.ZodArray
```
Is this the answer to your TS problem below?

Reply:
WOAH YEAH. Great catch here! The `schema instanceof z.ZodArray` was there to indicate that if the _previous_ schema was an array, the new result should also be an array. But the `newSchema` generation logic already has that logic baked in, so we can use `newSchema` in this check.